cpp/include/raft/neighbors/detail/ann_mg.cuh

```diff
@@ -0,0 +1,645 @@
+    } else if constexpr (std::is_same<AnnIndexType, ivf_pq::index<IdxT>>::value) {
+      return index_.value().size();
+    } else if constexpr (std::is_same<AnnIndexType, cagra::index<T, IdxT>>::value) {
+      return index_.value().size();
+    }
+  }
+
+ private:
+  std::optional<AnnIndexType> index_;
+};
+
+template <typename AnnIndexType, typename T, typename IdxT>
+class ann_mg_index {
+ public:
+  ann_mg_index(const std::vector<int>& dev_list, dist_mode mode = SHARDING)
+    : mode_(mode),
+      root_rank_(0),
+      num_ranks_(dev_list.size()),
+      dev_ids_(dev_list),
+      nccl_comms_(dev_list.size())
+  {
+    init_device_resources();
+    init_nccl_clique();
+  }
+
+  // deserialization
+  ann_mg_index(const raft::resources& handle,
+               const std::string& filename) {
+      std::ifstream is(filename, std::ios::in | std::ios::binary);
+      if (!is) { RAFT_FAIL("Cannot open file %s", filename.c_str()); }
+
+      mode_ = (raft::neighbors::mg::dist_mode)deserialize_scalar<int>(handle, is);
+      root_rank_ = 0;
+      num_ranks_ = deserialize_scalar<int>(handle, is);
+      dev_ids_.resize(num_ranks_);
+      std::iota(std::begin(dev_ids_), std::end(dev_ids_), 0);
+      nccl_comms_.resize(num_ranks_);
+
+      init_device_resources();
+      init_nccl_clique();
+
+      for (int rank = 0; rank < num_ranks_; rank++) {
+        RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+        auto& ann_if = ann_interfaces_.emplace_back();
+        ann_if.deserialize(dev_resources_[rank], is);
+      }
+
+      is.close();
+  }
+
+  ann_mg_index(const ann_mg_index&)                    = delete;
+  ann_mg_index(ann_mg_index&&)                         = default;
+  auto operator=(const ann_mg_index&) -> ann_mg_index& = delete;
+  auto operator=(ann_mg_index&&) -> ann_mg_index&      = default;
+
+  void init_device_resources() {
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+      dev_resources_.emplace_back();
+    }
+  }
+
+  void init_nccl_clique() {
+    RAFT_NCCL_TRY(ncclCommInitAll(nccl_comms_.data(), num_ranks_, dev_ids_.data()));
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+      raft::comms::build_comms_nccl_only(&dev_resources_[rank], nccl_comms_[rank], num_ranks_, rank);
+    }
+  }
+
+  void destroy_nccl_clique() {
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      cudaSetDevice(dev_ids_[rank]);
+      ncclCommDestroy(nccl_comms_[rank]);
+    }
+  }
+
+  void build(const ann::index_params* index_params,
+             raft::host_matrix_view<const T, IdxT, row_major> index_dataset)
+  {
+    if (mode_ == INDEX_DUPLICATION) {
+      for (int rank = 0; rank < num_ranks_; rank++) {
+        RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+        auto& ann_if = ann_interfaces_.emplace_back();
+        ann_if.build(dev_resources_[rank], index_params, index_dataset);
```

Comment:
Is it every GPU will copy the host dataset into device so the total number of copies will be num_ranks_?

Reply:
Another related question, will GPU 1 not start and wait until GPU 0 build finishes? If that's the case, the total runtime of the for loop seems to be single GPU build time * num_ranks_.

Reply:
Yes exactly, in the index duplication mode the dataset is copied in full to each GPU for training. An alternative method is to train a model locally, serialize it and distribute it with either one of the `distribute_flat`, `distribute_pq` or `distribute_cagra` functions.

> Another related question, will GPU 1 not start and wait until GPU 0 build finishes? If that's the case, the total runtime of the for loop seems to be single GPU build time * num_ranks_.

The `build`, `extend` and `search` functions take in a `handle` parameter containing the CUDA stream on which the kernels should be launched. This is supposed to be asynchronous and allow fast switching of GPUs. However, this has not yet been tested. An actual benchmark would be necessary to confirm that things scale as expected.

Reply:
We have a problem here. Building a GPU index is not only a GPU operation. It can have significant CPU work (e.g. CAGRA graph optimization, NN descent data pre/post proc, host side sub-sampling of IVF-methods).

Furthermore there are cases where our algorithms block CPU thread while waiting for GPU kernels to finish ( e.g. wait for return values that determine memory allocation size).

We cannot launch `build`  on a single CPU thread and expect that it will run parallel, because the GPU ops are asynchronous. Most are, but the the few that I cite above will essentially serialize the whole process.

At least we would need different worker threads for each GPU stream. But I would recommend one process per GPU.

We should also keep in mind that `build` is multi-threaded. It spawns OpenMP threads to help shuffle data in host memory (singe thread is not enough to saturate mem bandwidth). We should document that this can be controlled with the `OMP_NUM_THREADS` variable.